File: ABC/401-450/403/c.cpp
#include <iostream>
#include<vector>
#include<set>
using namespace std;

int main(){
  int n,m,q;
  cin >> n >> m  >> q;
  vector<set<int> > auth(n);
  vector<bool> ok(n);
  while(q--){
    int type;
    cin >> type;
    if(type == 1){
      int x,y;
      cin >> x >> y;
      x--; y--;
      auth[x].insert(y);
    }else if(type ==2){
      int x;
      cin >> x;
      x--;
      ok[x] = true;
    }else{
      int x,y;
      cin >> x >> y;
      x--; y--;
      if(ok[x]) cout << "Yes\n";
      else if(auth[x].count(y)) cout << "Yes\n";
      else cout << "No\n";
    }
  }
  return 0;
}

File: ABC/401-450/403/c.py
def main():
  n,m,q = map(int,input().split())
  done = [False]*n
  auth = [set() for _ in range(n)]
  for _ in range(q):
    txy = list(map(int,input().split()))
    t = txy[0]
    if t == 1:
      x,y = txy[1:]
      x-=1; y -= 1;
      auth[x].add(y)
    elif t == 2:
      x = txy[1]
      x -= 1
      done[x] = True
    else:
      x,y = txy[1:]
      x -=1; y -=1;
      if done[x]:
        print("Yes")
      elif y in auth[x]:
        print("Yes")
      else:
        print("No")

if __name__ == '__main__':
  main()


